src/object_accessor.hpp

```diff
@@ -246,23 +250,38 @@ Object Object::create(ContextType ctx, SharedRealm realm, const ObjectSchema &ob
         row_index = get_for_primary_key_impl(ctx, *table, *primary_prop, primary_value);
 
         if (row_index == realm::not_found) {
-            row_index = table->add_empty_row();
             created = true;
-            if (primary_prop->type == PropertyType::Int)
+            if (primary_prop->type == PropertyType::Int) {
+#if REALM_ENABLE_SYNC
+                row_index = sync::create_object_with_primary_key(realm->read_group(), *table, Accessor::to_long(ctx, primary_value));
+#else
+                row_index = table->add_empty_row();
                 table->set_int_unique(primary_prop->table_column, row_index, Accessor::to_long(ctx, primary_value));
+#endif // REALM_ENABLE_SYNC
+            }
             else if (primary_prop->type == PropertyType::String) {
                 auto value = Accessor::to_string(ctx, primary_value);
+#if REALM_ENABLE_SYNC
+                row_index = sync::create_object_with_primary_key(realm->read_group(), *table, value);
+#else
+                row_index = table->add_empty_row();
                 table->set_string_unique(primary_prop->table_column, row_index, value);
+#endif // REALM_ENABLE_SYNC
+            }
+            else {
+                REALM_TERMINATE("Unsupported primary key type.");
```

Comment:
`REALM_UNREACHABLE` is mainly intended to silence compiler warnings, I think — even though it does happen to abort in our version. I'm using `REALM_TERMINATE` as a substitute for `REALM_ASSERT(false)` — isn't it appropriate with the equivalent of an assertion here?